### PR TITLE
Add DuckDB QueryAssertions logic for MAP columns

### PR DESCRIPTION
**Pull request by aditi-pandit**
This allows us to write tests for MAP columns in Velox Operator
testing.

### CODE REVIEW

File: velox/exec/tests/LimitTest.cpp [hunk 1]
```diff
@@ -63,3 +63,43 @@ TEST_F(LimitTest, basic) {
...
+             [](auto row) { return 2; },
+             [](auto row) { return row % 2; },
+             [](auto row) { return row % 2 + 1; })}));
+  }
+  createDuckDbTable(vectors);
+  auto makePlan = [&](int32_t offset, int32_t limit) {
+    return PlanBuilder().values(vectors).limit(offset, limit, false).planNode();
+  };
+
+  assertQuery(makePlan(0, 10), "SELECT * FROM tmp LIMIT 10");
```
Comment (majetideepak) [inline velox/exec/tests/LimitTest.cpp:84]: Maybe put these `assertQuery`s in a function and reuse them with the `basic` test?

**Comment by mbasmanova** [comment]
@aditi-pandit Nice. I wonder if we can now streamline TEST_P(TableScanTest, constDictLazy) to remove the workaround:

```
  // Exclude map columns as DuckDB doesn't support complex types yet.
```

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 2]
```diff
@@ -175,6 +208,30 @@ velox::variant rowVariantAt(
   return velox::variant::row(std::move(values));
 }
 
+velox::variant mapVariantAt(
+    const ::duckdb::Value& vector,
+    const std::shared_ptr<const Type>& mapType) {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.cpp:213]: use TypePtr for readabiity

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 3]
```diff
@@ -175,6 +208,30 @@ velox::variant rowVariantAt(
...
+  std::map<variant, variant> map;
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.cpp:214]: nit: std::unordered_map

**Review by mbasmanova — COMMENTED** [review]
@aditi-pandit Nice clean change. The new test in LimitTest is a bit artificial. Consider using the new functionality in TEST_P(TableScanTest, constDictLazy) instead.

File: velox/exec/tests/utils/QueryAssertions.cpp (hunk 3 above)
Comment (aditi-pandit) [reply velox/exec/tests/utils/QueryAssertions.cpp:214]: There is some subtle interplay between unordered_map and variant causing compilation issues. Leaving this as map for now.

Comment (mbasmanova) [reply velox/exec/tests/utils/QueryAssertions.cpp:214]: Wow! Would you share any details?

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 4]
```diff
@@ -105,6 +108,36 @@ ::duckdb::Value duckValueAt<TypeKind::ARRAY>(
...
+  auto& mapKeys = mapVector->mapKeys();
+  auto& mapValues = mapVector->mapValues();
+  auto offset = mapVector->offsetAt(row);
+  auto size = mapVector->sizeAt(row);
+
+  std::vector<::duckdb::Value> duckKeysVector;
+  std::vector<::duckdb::Value> duckValuesVector;
+  duckKeysVector.reserve(size);
+  duckValuesVector.reserve(size);
+  for (auto i = 0; i < size; i++) {
```
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/exec/tests/utils/QueryAssertions.cpp:125]: I'm going to pass on this one as the common function is limited reuse right now. The current code clearly reads into what is happening.

**Comment by aditi-pandit** [comment]
> @aditi-pandit Nice clean change. The new test in LimitTest is a bit artificial. Consider using the new functionality in TEST_P(TableScanTest, constDictLazy) instead.

Done.

File: velox/exec/tests/utils/QueryAssertions.cpp (hunk 3 above)
Comment (aditi-pandit) [reply velox/exec/tests/utils/QueryAssertions.cpp:214]: In template: call to implicitly-deleted default constructor of 'std::hash<facebook::velox::variant>' error occurred here in instantiation of member function 'std::__unordered_map_hasher<facebook::velox::variant, std::__hash_value_type<facebook::velox::variant, facebook::velox::variant>, std::hash<facebook::velox::variant>, std::equal_to<facebook::velox::v... in instantiation of member function 'std::__hash_table<std::__hash_value_type<facebook::velox::variant, facebook::velox::variant>, std::__unordered_map_hasher<facebook::velox::variant, std::__hash_value_type<facebook::velox::variant, fa... in instantiation of member function 'std::unordered_map<facebook::velox::variant, facebook::velox::variant>::unordered_map' requested here '__enum_hash' has been explicitly marked deleted here

File: velox/exec/tests/TableScanTest.cpp [hunk 5]
```diff
@@ -333,16 +333,14 @@ TEST_P(TableScanTest, constDictLazy) {
        makeFlatVector<int64_t>(size, [](auto row) { return row; }),
        makeMapVector<int64_t, double>(
            size,
-           [](auto row) { return row % 3; },
+           [](auto row) { return row % 3 + 1; },
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:336]: Why is this change? Can you not write empty maps into DuckDB? If that's the case, let's file a GitHub issue with DuckDB.

Comment (aditi-pandit) [reply velox/exec/tests/TableScanTest.cpp:336]: DuckDB requires use of the EMPTYLIST API which needs the DuckDB LogicalType for each Velox type. I didn't find any straightforward way to do so in Velox quite yet.

//! Create an empty list with the specified child-type
DUCKDB_API static Value EMPTYLIST(LogicalType child_type);

I didn't see the use of EMPTYLIST in Array etc either.

Are you familiar with anything for this ? I'm unsure if this is a gap on our side or theirs.

Comment (mbasmanova) [reply velox/exec/tests/TableScanTest.cpp:336]: Hmm... I can take a look tomorrow. @Mytherin Mark, would you know this off the top of your head?

Comment (mbasmanova) [reply velox/exec/tests/TableScanTest.cpp:336]: @aditi-pandit I think you can use fromVeloxType from velox/duckdb/conversion/DuckConversion.h to get LogicalType and pass that to Value::EMPTYLIST.

Comment (aditi-pandit) [reply velox/exec/tests/TableScanTest.cpp:336]: This has only scalar types. But that's good enough for now. Will update the code.

File: velox/exec/tests/TableScanTest.cpp [hunk 6]
```diff
@@ -340,9 +340,7 @@ TEST_P(TableScanTest, constDictLazy) {
   auto filePath = TempFilePath::create();
   writeToFile(filePath->path, kTableScanTest, {rowVector});
 
-  // Exclude map columns as DuckDB doesn't support complex types yet.
-  createDuckDbTable(
-      {makeRowVector({rowVector->childAt(0), rowVector->childAt(1)})});
+  createDuckDbTable({rowVector});
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:343]: Nice and clean. Thank you, Aditi.

**Review by mbasmanova — APPROVED** [review]
@aditi-pandit Thanks. The changes in LimitTest seem unnecessary. They complicate the test, but do not seem to increase coverage. Am I missing something?

**Comment by aditi-pandit** [comment]
> @aditi-pandit Thanks. The changes in LimitTest seem unnecessary. They complicate the test, but do not seem to increase coverage. Am I missing something?

HI Masha,

  The LimitTest covers both parts : i) Inserting MAP column values from Velox -> DuckDB and ii) Retrieving DuckDB Map columns into Velox map structures. 

  My understanding of the TableScan test was that it covered insertion logic but not retrieving part.

  Please let me know otherwise.

**Comment by aditi-pandit** [comment]
#1066

**Pull request closed without merging**